<div class="alert alert-block alert-success">

# <center>  </center>
# <center> Amazon Appliances' Reviews </center>
### <center> Extract Reviews and Metadata</center>
#### <center> JSON/CSV files</center>

**Author: Atef Bader, PhD**
    
**Last Edit: 2/15/2024**  

</div>

In [16]:
import json
import ast
from pandas import DataFrame, Series      


In [17]:

fr=open("meta_Appliances.json")
fw=open("Amazon_Appliances_Metadata.json", "w")

for line in fr:
    json_dat = json.dumps(ast.literal_eval(line))
    dict_dat = json.loads(json_dat)
    json.dump(dict_dat, fw)
    fw.write("\n")

fw.close()
fr.close()


In [18]:

fr=open("reviews_Appliances.json")
fw=open("Amazon_Appliances_Reviews.json", "w")

for line in fr:
    json_dat = json.dumps(ast.literal_eval(line))
    dict_dat = json.loads(json_dat)
    json.dump(dict_dat, fw)
    fw.write("\n")

fw.close()
fr.close()


In [19]:

list_of_appliances_metadata_dict_data = [json.loads(line) for line in open('Amazon_Appliances_Metadata.json')]

In [20]:
appliances_metadata_df = DataFrame(list_of_appliances_metadata_dict_data)

In [21]:
appliances_metadata_df.head()

,asin,price,imUrl,description,categories,title,brand,related,salesRank
0,0970408285,40.75,http://ecx.images-amazon.com/images/I/41kQjPIF...,High-pressure hose 90 end & hose washer instal...,"[[Appliances, Parts & Accessories, Washer Part...",NaN,NaN,NaN,NaN
1,7301113188,NaN,http://ecx.images-amazon.com/images/I/41EfFOPw...,NaN,"[[Appliances, Freezers, Chest Freezers]]",NaN,NaN,NaN,NaN
2,B00002N7HY,2.29,http://ecx.images-amazon.com/images/I/316Y1KbT...,"Built of durable thermoplastic, Leviton Power ...","[[Appliances, Parts & Accessories, Range Parts...","Leviton 5050 50 Amp, 125/250 Volt, NEMA 10-50R...",Leviton,"{'also_bought': ['B000BPFZHM', 'B000VU1KC6', '...",NaN
3,B00002NATH,9.13,http://ecx.images-amazon.com/images/I/41srC2QF...,"Built of durable thermoplastic, Leviton Power ...","[[Appliances, Parts & Accessories, Range Parts...","Leviton 7313 30 Amp, 125 Volt, NEMA Tt-30R, 2P...",Leviton,"{'also_bought': ['B002OUMU66', 'B007HFT034', '...",NaN
4,B00004SQH9,16.29,http://ecx.images-amazon.com/images/I/41eg0Hvy...,Coleman Cable 3536 14/3 General-Use Appliance ...,"[[Appliances, Parts & Accessories, Refrigerato...",Coleman Cable 3536 14/3 General-Use Appliance ...,Coleman Cable,"{'also_bought': ['B000XSNDS0', 'B0022NH420', '...",NaN


In [22]:

list_of_appliances_reviews_dict_data = [json.loads(line) for line in open('Amazon_Appliances_Reviews.json')]

In [23]:
appliances_reviews_df = DataFrame(list_of_appliances_reviews_dict_data)

In [24]:
appliances_reviews_df.head()

,reviewerID,asin,reviewerName,helpful,reviewText,overall,summary,unixReviewTime,reviewTime
0,A27HOWZBUBJ8FF,0970408285,Steve,"[0, 0]",Could have been longer though. well made and e...,4.0,Good fit,1387152000,"12 16, 2013"
1,A24HQ894NFSTF5,7301113188,"Maha Saqfalhait ""shopaholic! ;)""","[0, 0]",I like these containers so much i have ordered...,5.0,I Love the Freezer storage line..,1236902400,"03 13, 2009"
2,AXE83MK90ZEVZ,B00002N7HY,Strom,"[0, 0]","It works, no fires, etc. Why not 5 stars? Ho...",4.0,expectations achieved.,1389052800,"01 7, 2014"
3,A2J7X7ZIH2EWB1,B00002NATH,NaN,"[0, 0]",Fast shipping. Works great,5.0,Five Stars,1405814400,"07 20, 2014"
4,AJQFNOFTZ7GOX,B00002NATH,Barthbill,"[1, 1]",What can I say? It is the usual Leviton high q...,5.0,good product at a good price.,1277164800,"06 22, 2010"


In [25]:
appliances_reviews_df.to_csv('Amazon_Appliances_Reviews.csv', encoding='utf-8', index=False)

In [26]:
appliances_metadata_df.to_csv('Amazon_Appliances_Metadata.csv', encoding='utf-8', index=False)

<div class="alert alert-block alert-danger">

**Requirement 1:** Requirement 1: (10 Points) Inspect the provided datasets and run the
provided Extract_Amazon_Appliances_Reviews_data.ipynb.script.
Comment on the different fields (data types, NaN values, etc.) in the
created CSV files (Amazon_Appliances_Metadata.csv and
Amazon_Appliances_Reviews.csv ) and what data preprocessing is
needed to prepare the data to be loaded into a database. The asin - ID
of the product is the common field between the two tables, do you think
that is sufficient to cross-reference the data entries/rows in the two
datasets? Explain.

</div>

In [27]:
appliances_metadata_df.info()
appliances_metadata_df.describe()

 # Check for missing values in key columns
asin_null = appliances_metadata_df['asin'].isnull().sum()
title_null = appliances_metadata_df['title'].isnull().sum()
description_null = appliances_metadata_df['description'].isnull().sum()
imUrl_null = appliances_metadata_df['imUrl'].isnull().sum()

print(f"\nMissing values:")
print(f"asin values: {asin_null}")
print(f"title values: {title_null}")
print(f"Description null values: {description_null}")
print(f"imUrl null values: {imUrl_null}")

# Check 1:1 relationship - group by asin and imurls
asin_to_imUrl_mapping = appliances_metadata_df.groupby('asin')['imUrl'].nunique()
multiple_asin_imUrl_desc = (asin_to_imUrl_mapping > 1).sum()
print(f"asin with multiple imUrl: {multiple_asin_imUrl_desc}")

# Check 1:1 relationship - group by asin and descriptions
asin_to_description_mapping = appliances_metadata_df.groupby('asin')['description'].nunique()
multiple_asin_desciption_desc = (asin_to_description_mapping > 1).sum()
print(f"asin with multiple descriptions: {multiple_asin_desciption_desc}")

# Check one cent items
price_001_df = appliances_metadata_df[appliances_metadata_df['price'] == 0.01]
print(f"Records with price = 0.01: {len(price_001_df)}")


# Filter out rows with null descriptions or URLs for the relationship analysis
valid_data = appliances_metadata_df.dropna(subset=['title', 'description', 'imUrl'])
print(f"Records with title, description and imUrl: {len(valid_data)}")



price_001_df.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11656 entries, 0 to 11655
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   asin         11656 non-null  object 
 1   price        8098 non-null   float64
 2   imUrl        11647 non-null  object 
 3   description  11134 non-null  object 
 4   categories   11656 non-null  object 
 5   title        6635 non-null   object 
 6   brand        4078 non-null   object 
 7   related      9070 non-null   object 
 8   salesRank    2639 non-null   object 
dtypes: float64(1), object(8)
memory usage: 819.7+ KB

Missing values:
asin values: 0
title values: 5021
Description null values: 522
imUrl null values: 9
asin with multiple imUrl: 0
asin with multiple descriptions: 0
Records with price = 0.01: 9
Records with title, description and imUrl: 6465


,asin,price,imUrl,description,categories,title,brand,related,salesRank
962,B000KKISE0,0.01,http://ecx.images-amazon.com/images/I/51ox8JBs...,"Top In dimensions: 7.5"".",[[Appliances]],NaN,NaN,"{'also_bought': ['B00DM8L2AO', 'B0026KSN92', '...",NaN
2090,B001B2VXPW,0.01,http://ecx.images-amazon.com/images/I/61XL24bT...,"Aluminum burner bibs which are disposable, str...","[[Appliances, Parts & Accessories, Range Parts...",Pactiv/E Z Foil 88801A Gas Burner Bib For Stove,NaN,"{'also_bought': ['B000KKISE0'], 'buy_after_vie...",NaN
2250,B001DHLPYE,0.01,http://ecx.images-amazon.com/images/I/01O-iguZ...,Genuine OEM Dryer Drum Bearing Slide. Only 1 p...,"[[Appliances, Parts & Accessories, Dryer Parts...",NaN,NaN,"{'also_bought': ['B003BVPY5G', 'B004H3XU3K', '...",NaN
2286,B001DHLQT8,0.01,http://ecx.images-amazon.com/images/I/21oiLZ5u...,"Whirlpool Part Number 2194719: Retainer, Levers","[[Appliances, Parts & Accessories]]",NaN,NaN,"{'also_bought': ['B00DM8L0AG', 'B0053FA098', '...",NaN
3186,B00202PEZ2,0.01,http://ecx.images-amazon.com/images/I/31n-3Azt...,"Latch and seal kit used on Maytag, Amana, Jenn...","[[Appliances, Parts & Accessories, Dishwasher ...",NaN,NaN,"{'also_viewed': ['B00DM8KJSA', 'B005AR74L6', '...",NaN
3997,B002ZNC2BU,0.01,http://ecx.images-amazon.com/images/I/11vSG072...,Whirlpool Part Number 842177: Key,"[[Appliances, Parts & Accessories]]",NaN,NaN,"{'also_viewed': ['B002ZNICPU', 'B004X85N2A', '...",NaN
5913,B004MNF0S8,0.01,http://ecx.images-amazon.com/images/I/51zfIRqE...,Emerson Appliance Solutions Mp15ya-404080 Rang...,"[[Appliances, Parts & Accessories, Range Parts...",NaN,NaN,"{'also_bought': ['B002ZNNE12', 'B000FNJ5RG', '...",NaN
7366,B005CY0G1M,0.01,http://ecx.images-amazon.com/images/I/31w%2Bnw...,DUPLICATE: Broan Kenmore Hood Light Lens 99110437,"[[Appliances, Parts & Accessories]]",NaN,NaN,"{'also_viewed': ['B00GVKIYHE', 'B000IBEYOO', '...",NaN
11328,B00GQ12AC8,0.01,http://ecx.images-amazon.com/images/I/415xXaBM...,Part Number WD12X10186 (AP3854206) replaces WD...,"[[Appliances, Parts & Accessories, Range Parts...",(4 PACK) PART # WD12X10186 GENUINE FACTORY OEM...,NaN,"{'also_viewed': ['B00917EZPO', 'B00917D6TA', '...",NaN


For the meta data df, there are a total of 9 columns with 11656 entries. There are quite a few missing values issues that needs to be sorted out with
brand, title, price, salesrank leading the ranks of missing values. salesrank information may be worth droping as only of 18% of asin contain this information.
Just examinging 0.01 cent price information identified 9 items that clears costs more than 1 cent so accuracy of pricing is questionable. Maybe need additional qualtity control of pricing information thorugh other sources of data. Good news is there were no multiple matches of asin to imUrl or description. 

In [28]:
appliances_reviews_df.info()
appliances_reviews_df.describe()


# Filter out rows with null reviewerID or asin for the relationship analysis
valid_review_data = appliances_reviews_df.dropna(subset=['reviewerID', 'asin'])
print(f"Records with both reviewerID and asin: {len(valid_review_data)}")



import pandas as pd
from datetime import datetime

# Convert unixReviewTime to datetime
appliances_reviews_df['unixReviewTime_converted'] = pd.to_datetime(appliances_reviews_df['unixReviewTime'], unit='s')

# Convert reviewTime string to datetime
appliances_reviews_df['reviewTime_converted'] = pd.to_datetime(appliances_reviews_df['reviewTime'], format='%m %d, %Y')

# Show non-matching rows if any
non_matching = appliances_reviews_df[appliances_reviews_df['unixReviewTime_converted'].dt.date != appliances_reviews_df['reviewTime_converted'].dt.date]
print(f"Non-matching rows: {len(non_matching)}")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 143685 entries, 0 to 143684
Data columns (total 9 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   reviewerID      143685 non-null  object 
 1   asin            143685 non-null  object 
 2   reviewerName    143044 non-null  object 
 3   helpful         143685 non-null  object 
 4   reviewText      143685 non-null  object 
 5   overall         143685 non-null  float64
 6   summary         143685 non-null  object 
 7   unixReviewTime  143685 non-null  int64  
 8   reviewTime      143685 non-null  object 
dtypes: float64(1), int64(1), object(7)
memory usage: 9.9+ MB
Records with both reviewerID and asin: 143685
Non-matching rows: 0


For the reviews df, only missing data detected was with reviewer name column which is not that critical for the purposes of sentiment review.
Checking for Duplicate reviews by same reviewer did not identiy any. 

In [29]:
# Analysis of ASIN relationship between metadata and reviews DataFrames
print("=== ASIN Analysis: Metadata vs Reviews ===\n")

# Get unique ASINs from each DataFrame
metadata_asins = set(appliances_metadata_df['asin'].dropna())
reviews_asins = set(appliances_reviews_df['asin'].dropna())

print(f"Unique ASINs in metadata: {len(metadata_asins)}")
print(f"Unique ASINs in reviews: {len(reviews_asins)}")

# Check overlap
common_asins = metadata_asins.intersection(reviews_asins)
metadata_only = metadata_asins - reviews_asins
reviews_only = reviews_asins - metadata_asins

print(f"\nASIN overlap analysis:")
print(f"ASINs in both datasets: {len(common_asins)}")
print(f"ASINs only in metadata: {len(metadata_only)}")
print(f"ASINs only in reviews: {len(reviews_only)}")

# 1:1 relationship check
is_perfect_match = len(metadata_asins) == len(reviews_asins) == len(common_asins)
print(f"\nPerfect 1:1 ASIN match: {is_perfect_match}")

# Coverage percentages
metadata_coverage = len(common_asins) / len(metadata_asins) * 100 if metadata_asins else 0
reviews_coverage = len(common_asins) / len(reviews_asins) * 100 if reviews_asins else 0

print(f"\nCoverage analysis:")
print(f"Metadata ASINs with reviews: {metadata_coverage:.1f}%")
print(f"Review ASINs with metadata: {reviews_coverage:.1f}%")

# Show examples of non-matching ASINs
if metadata_only:
    print(f"\nExample ASINs only in metadata (first 5): {list(metadata_only)[:5]}")
if reviews_only:
    print(f"Example ASINs only in reviews (first 5): {list(reviews_only)[:5]}")

print(f"\n=== SUMMARY ===")
print(f"Total unique ASINs across both datasets: {len(metadata_asins.union(reviews_asins))}")
print(f"Relationship type: {'Perfect 1:1 match' if is_perfect_match else 'Not 1:1 - datasets have different ASIN coverage'}")

=== ASIN Analysis: Metadata vs Reviews ===

Unique ASINs in metadata: 11656
Unique ASINs in reviews: 11402

ASIN overlap analysis:
ASINs in both datasets: 11402
ASINs only in metadata: 254
ASINs only in reviews: 0

Perfect 1:1 ASIN match: False

Coverage analysis:
Metadata ASINs with reviews: 97.8%
Review ASINs with metadata: 100.0%

Example ASINs only in metadata (first 5): ['B0050KJZME', 'B00ECX09BG', 'B009U1NQEM', 'B0000SWBM2', 'B0002UMYDO']

=== SUMMARY ===
Total unique ASINs across both datasets: 11656
Relationship type: Not 1:1 - datasets have different ASIN coverage


In terms of cross-referencing and assessing whether or not asin is sufficient as primary key, we identified that out of 14687 records found in the reviewer df, only 11402 asin were detected in both df (2.3% missing meta data). However, with 11402 records found in both df, there were no duplicates so we are good to use asin for these records.

<div class="alert alert-block alert-danger">

**Requirement 2:** (10 Points) Create a table for a sample list of 3 queries
(NOT listed in Requirements Specification Document) for every category
listed below:
1. Quantitative Queries
2. Fuzzy/Exact-Search Queries
3. Hybrid-Search Queries

</div>

In [ ]:
queries_data = {
    'Quantitative Queries': [
        'What are the 25th, 50th (median), 75th, and 95th percentiles of product prices for each product subcategory that has at least 10 products?',
        'What is the correlation between review length and helpfulness ratio for each product subcategory with at least 50 reviews?',
        'Which brands show the strongest seasonal performance variation (highest difference between best and worst quarterly average ratings)for brands with at least 100 reviews?'
    ]
}

queries_df = pd.DataFrame(queries_data)

print(queries_df.to_string(index=False))

                                                                                                                                                      Quantitative Queries
                                What are the 25th, 50th (median), 75th, and 95th percentiles of product prices for each product subcategory that has at least 10 products?
                                                What is the correlation between review length and helpfulness ratio for each product subcategory with at least 50 reviews?
Which brands show the strongest seasonal performance variation (highest difference between best and worst quarterly average ratings) for brands with at least 100 reviews?


In [31]:
# Create three additional fuzzy matching queries following the examples pattern
fuzzy_queries_data = {
    'Fuzzy Matching Queries': [
        'Get a list of those reviews where the same reviewer complained about loud noise for washing machine but praised the quiet operation for microwave?',
        'Get a list of those reviews where the same reviewer complained about installation difficulty for range but found installation easy for garbage disposal?',
        'Get a list of those reviews where the reviewer mentioned excellent performance but also complained about poor build quality for the same appliance?'
    ]
}

# Create DataFrame
fuzzy_queries_df = pd.DataFrame(fuzzy_queries_data)

# Display the table
print("Additional Fuzzy Matching Queries:")
print("=" * 50)
print(fuzzy_queries_df.to_string(index=False))

Additional Fuzzy Matching Queries:
                                                                                                                                  Fuzzy Matching Queries
      Get a list of those reviews where the same reviewer complained about loud noise for washing machine but praised the quiet operation for microwave?
Get a list of those reviews where the same reviewer complained about installation difficulty for range but found installation easy for garbage disposal?
     Get a list of those reviews where the reviewer mentioned excellent performance but also complained about poor build quality for the same appliance?


In [32]:
hybrid_queries_data = {
    'Hybrid Queries': [
        'Get a list of those reviews that are similar to this text: "The installation was a nightmare, took hours to figure out the connections and the manual was useless" and were reviewed for range/oven products with ratings below 3 stars.',
        'Get a list of those reviews that are similar to this text: "Great value for money but the build quality feels cheap and flimsy" and got reviewed between January 2012 and December 2013.',
        'Get a list of those reviews that are similar to this text: "The customer service was terrible, no help at all when I called about warranty issues" and are for products in the dishwasher category with helpful votes greater than 5.'
    ]
}

# Create DataFrame
hybrid_queries_df = pd.DataFrame(hybrid_queries_data)

# Display the table
print("Additional Hybrid Queries:")
print("=" * 50)
print(hybrid_queries_df.to_string(index=False))

Additional Hybrid Queries:
                                                                                                                                                                                                                          Hybrid Queries
Get a list of those reviews that are similar to this text: "The installation was a nightmare, took hours to figure out the connections and the manual was useless" and were reviewed for range/oven products with ratings below 3 stars.
                                                Get a list of those reviews that are similar to this text: "Great value for money but the build quality feels cheap and flimsy" and got reviewed between January 2012 and December 2013.
   Get a list of those reviews that are similar to this text: "The customer service was terrible, no help at all when I called about warranty issues" and are for products in the dishwasher category with helpful votes greater than 5.


<div class="alert alert-block alert-danger">

***Requirement 3:*** (10 Points) Use ChatGPT and enter the CoT text (Section 5,
Page 18 of Final Project - Requirements Specification.pdf) into ChatGPT
prompt to get the recommended database that best meets the needs of
every query you identified above. Create a table that has two columns:
the query and the recommended database for it as printed by ChatGPT.

</div>

In [46]:
from openai import OpenAI
import os
import json
import requests
from dotenv import load_dotenv  # Add this import
from tenacity import retry, wait_random_exponential, stop_after_attempt
from termcolor import colored

# Load environment variables
load_dotenv()  

# Updated model 
GPT_MODEL = "gpt-3.5-turbo"

# SECURE: Use environment variable
client = OpenAI(
    api_key=os.getenv("OPENAI_API_KEY")
)

@retry(wait=wait_random_exponential(min=1, max=60), stop=stop_after_attempt(6))
def get_completion(prompt, model=GPT_MODEL, temperature=0):
    """Get completion with retry logic and error handling"""
    try:
        messages = [{"role": "user", "content": prompt}]
        response = client.chat.completions.create(
            model=model,
            messages=messages,
            temperature=temperature,
        )
        return response.choices[0].message.content
    except Exception as e:
        print(f"Error: {e}")
        return None

# Test the function
def test_connection():
    response = get_completion("Say this is a test")
    print(response)

# Run the test
test_connection()



This is a test.


In [47]:
# Extract the questions from column 0 of each DataFrame
quantitative_questions = queries_df.iloc[:, 0].tolist()
fuzzy_questions = fuzzy_queries_df.iloc[:, 0].tolist()
hybrid_questions = hybrid_queries_df.iloc[:, 0].tolist()

# Combine all questions into one list
all_questions = quantitative_questions + fuzzy_questions + hybrid_questions

# Create a single DataFrame with all questions
combined_queries_df = pd.DataFrame({'All Queries': all_questions})

# Display the combined table
print("All Queries Combined:")
print("=" * 60)
print(combined_queries_df.to_string(index=False))

All Queries Combined:
                                                                                                                                                                                                                             All Queries
                                                                                              What are the 25th, 50th (median), 75th, and 95th percentiles of product prices for each product subcategory that has at least 10 products?
                                                                                                              What is the correlation between review length and helpfulness ratio for each product subcategory with at least 50 reviews?
                                                              Which brands show the strongest seasonal performance variation (highest difference between best and worst quarterly average ratings) for brands with at least 100 reviews?
                                              

In [48]:
text = """You are a database application developer.
Follow these steps to answer the user queries:
Step 1: Parse the user query to determine the type of the query and
expected output/result of the query.
Step 2: Determine the type of the required database system by
performing the following steps.
Step 2.1 if the query asks for results in the form of numerical values,
calculation, joining multiple tables, etc. then use Postgres.
Step 2.2 if the query asks for exact/fuzzy search of strings in
documents, then use ELasticSearch.
Step 2.3 if the query asks for similarity/semantic/hybrid search, then
use Milvus.
Step 3: Use the database system you identified in Step 2, create a query
to execute using that database
I have Amazon product reviews dataset. Which database system I should
use for this type of query: What is the average price of every product
subcategory that has an average overall rating that is greater than 3?
I have Amazon product reviews dataset. Which database system I should
use for this type of query: Get a list of those reviews where the reviewer
reported a problem about the ice maker for a refrigerator with French
doors?
I have Amazon product reviews dataset. Which database system I should
use for this type of query: Get a list of those reviews that are similar to this
text: “This microwave started to make a lot of noise after using it for only
few days” and got reviewed in February of 2013.
"""




In [49]:
from IPython.display import display, HTML

# Join all queries into a single string
queries_text = "\n".join(combined_queries_df['All Queries'].tolist())

prompt = f"""
Following the Chain of Thought reasoning approach shown in the example text delimited by triple quotes, recommend the best database type for each query.
Example Chain of Thought is provided in:
  ```{text}```

Now apply this same reasoning approach to ALL {len(combined_queries_df)} queries below:

{queries_text}
```{text}```

Create a table with columns: "Query", "Database Type"
Include ALL queries with your chain of thought reasoning for each recommendation.
"""

response = get_completion(prompt)
print(response)


1. What are the 25th, 50th (median), 75th, and 95th percentiles of product prices for each product subcategory that has at least 10 products?
- Query Type: Statistical calculation on numerical values
- Database Type: Postgres

2. What is the correlation between review length and helpfulness ratio for each product subcategory with at least 50 reviews?
- Query Type: Statistical calculation on numerical values
- Database Type: Postgres

3. Which brands show the strongest seasonal performance variation (highest difference between best and worst quarterly average ratings) for brands with at least 100 reviews?
- Query Type: Statistical calculation on numerical values
- Database Type: Postgres

4. Get a list of those reviews where the same reviewer complained about loud noise for washing machine but praised the quiet operation for microwave?
- Query Type: Exact search of strings in documents
- Database Type: ELasticSearch

5. Get a list of those reviews where the same reviewer complained abou

<div class="alert alert-block alert-danger">

***Requirement 4:***  (10 Points) Use ChatGPT to generate code for every query
you identified above using the recommended database. Please see
sample prompts and output in Appendix A of this document.

</div>

In [50]:
example_postgres = f"""I have Amazon product reviews dataset. Use Postgres database system and
create code for the query: Get a list of review counts for every product category
in every month of the year 2013."""


prompt = f"""
I have Amazon product reviews dataset. Use Postgres database system created from appliances_metadata_df, and appliances_reviews_df to 
create code for the query: What are the 25th, 50th (median), 75th, and 95th percentiles of product prices for each product subcategory that has at least 10 products?
"""


response = get_completion(prompt)
print(response)

To achieve this, you can use the following SQL query:

```sql
WITH product_prices AS (
    SELECT 
        subcategory,
        price,
        PERCENT_RANK() OVER (PARTITION BY subcategory ORDER BY price) AS percentile
    FROM appliances_metadata_df
    WHERE subcategory IN (
        SELECT subcategory
        FROM appliances_metadata_df
        GROUP BY subcategory
        HAVING COUNT(*) >= 10
    )
)
SELECT 
    subcategory,
    PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY price) AS percentile_25th,
    PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY price) AS median,
    PERCENTILE_CONT(0.75) WITHIN GROUP (ORDER BY price) AS percentile_75th,
    PERCENTILE_CONT(0.95) WITHIN GROUP (ORDER BY price) AS percentile_95th
FROM product_prices
GROUP BY subcategory;
```

This query first filters out the product subcategories that have at least 10 products. Then, it calculates the 25th, 50th (median), 75th, and 95th percentiles of product prices for each of these subcategories using the `PERCENT

In [51]:
example_postgres = f"""I have Amazon product reviews dataset. Use Postgres database system and
create code for the query: Get a list of review counts for every product category
in every month of the year 2013."""


prompt = f"""
I have Amazon product reviews dataset. Use Postgres database system created from appliances_metadata_df, and appliances_reviews_df. 
Create code for the query: 
Which brands show the strongest seasonal performance variation (highest difference between best 
and worst quarterly average ratings) for brands with at least 100 reviews?
"""

response = get_completion(prompt)
print(response)

To find the brands with the strongest seasonal performance variation (highest difference between best and worst quarterly average ratings) for brands with at least 100 reviews, you can use the following SQL query:

```sql
WITH brand_avg_ratings AS (
    SELECT 
        brand,
        EXTRACT(quarter FROM review_date) AS quarter,
        AVG(rating) AS avg_rating
    FROM appliances_reviews_df
    GROUP BY brand, EXTRACT(quarter FROM review_date)
),
brand_quarterly_stats AS (
    SELECT
        brand,
        MAX(avg_rating) AS max_avg_rating,
        MIN(avg_rating) AS min_avg_rating
    FROM brand_avg_ratings
    GROUP BY brand
    HAVING COUNT(*) >= 4 -- At least 4 quarters of data
),
brand_seasonal_variation AS (
    SELECT
        brand,
        MAX(max_avg_rating) - MIN(min_avg_rating) AS seasonal_variation
    FROM brand_quarterly_stats
    GROUP BY brand
    HAVING COUNT(*) >= 4 -- At least 4 quarters of data
)
SELECT
    brand,
    seasonal_variation
FROM brand_seasonal_variati

In [52]:
example_postgres = f"""I have Amazon product reviews dataset. Use Postgres database system and
create code for the query: Get a list of review counts for every product category
in every month of the year 2013."""


prompt = f"""
I have Amazon product reviews dataset. Use Postgres database system created from appliances_metadata_df, and appliances_reviews_df. 
Create code for the query: What is the correlation between review length and helpfulness ratio for each product subcategory with at least 50 reviews?
"""

response = get_completion(prompt)
print(response)

To calculate the correlation between review length and helpfulness ratio for each product subcategory with at least 50 reviews, you can use the following SQL query:

```sql
WITH review_data AS (
    SELECT 
        a.product_id,
        a.product_category,
        a.review_length,
        b.helpful_votes / b.total_votes AS helpfulness_ratio
    FROM appliances_metadata_df a
    JOIN appliances_reviews_df b ON a.product_id = b.product_id
    WHERE a.product_category IN (
        SELECT product_category
        FROM appliances_metadata_df
        GROUP BY product_category
        HAVING COUNT(*) >= 50
    )
)

SELECT 
    product_category,
    product_id,
    CORR(review_length, helpfulness_ratio) AS correlation
FROM review_data
GROUP BY product_category, product_id
ORDER BY product_category, correlation DESC;
```

This query first creates a common table expression (CTE) `review_data` that combines the necessary data from the `appliances_metadata_df` and `appliances_reviews_df` tables. I

In [53]:
example_elastic_search = f"""I have Amazon product reviews dataset. Use ElasticSearch database system
and create code for the query: Get a list of those reviews that are similar to this
text: “This microwave started to make a lot of noise after using it for only few
days” and got reviewed in February of 2013 but the review misspelled
microwave"""


prompt = f"""
I have Amazon product reviews dataset. Use ElasticSearch database system
and create code for the query: Get a list of those reviews that are similar to this
text: “Get a list of those reviews where the same reviewer complained about loud noise for washing machine but praised the quiet operation for microwave?"
"""

response = get_completion(prompt)
print(response)




To achieve this, you can use the following code to query the Amazon product reviews dataset in ElasticSearch:

```python
from elasticsearch import Elasticsearch

# Connect to ElasticSearch
es = Elasticsearch()

# Define the search query
query = {
    "query": {
        "bool": {
            "must": [
                {"match": {"reviewer_name": "John Doe"}},  # Change to the actual reviewer name
                {"match": {"review_text": "loud noise washing machine"}}
            ],
            "should": [
                {"match": {"review_text": "quiet operation microwave"}}
            ]
        }
    }
}

# Execute the search query
results = es.search(index="amazon_reviews", body=query)

# Print the results
for hit in results['hits']['hits']:
    print(hit['_source'])
```

In this code snippet, we are connecting to the ElasticSearch database and defining a search query that looks for reviews where the reviewer complained about loud noise for a washing machine and praised the quiet op

In [54]:
example_elastic_search = f"""I have Amazon product reviews dataset. Use ElasticSearch database system
and create code for the query: Get a list of those reviews that are similar to this
text: “This microwave started to make a lot of noise after using it for only few
days” and got reviewed in February of 2013 but the review misspelled
microwave"""


prompt = f"""
I have Amazon product reviews dataset. Use ElasticSearch database system
and create code for the query: Get a list of those reviews that are similar to this
text: Get a list of those reviews where the same reviewer complained about installation difficulty for range but found installation easy for garbage disposal?
"""

response = get_completion(prompt)
print(response)

To achieve this, you can use the following code to query the ElasticSearch database:

```python
from elasticsearch import Elasticsearch

# Connect to ElasticSearch
es = Elasticsearch()

# Define the search query
query = {
    "query": {
        "bool": {
            "must": [
                {"match": {"review_text": "installation difficulty for range"}},
                {"match": {"review_text": "installation easy for garbage disposal"}}
            ]
        }
    }
}

# Execute the search query
results = es.search(index="amazon_reviews", body=query)

# Extract and print the list of reviews
for hit in results['hits']['hits']:
    print(hit['_source']['review_text'])
```

Make sure to replace "amazon_reviews" with the actual index name where your Amazon product reviews dataset is stored in ElasticSearch. This code will search for reviews where the same reviewer complained about installation difficulty for a range but found installation easy for a garbage disposal.


In [55]:
example_elastic_search = f"""I have Amazon product reviews dataset. Use ElasticSearch database system
and create code for the query: Get a list of those reviews that are similar to this
text: “This microwave started to make a lot of noise after using it for only few
days” and got reviewed in February of 2013 but the review misspelled
microwave"""


prompt = f"""
I have Amazon product reviews dataset. Use ElasticSearch database system
and create code for the query: Get a list of those reviews that are similar to this
text: "Get a list of those reviews where the reviewer mentioned excellent performance but also complained about poor build quality for the same appliance?"
"""

response = get_completion(prompt)
print(response)

To achieve this, you can use the following code to query the Amazon product reviews dataset in ElasticSearch:

```json
{
  "query": {
    "bool": {
      "must": [
        { "match": { "review_text": "excellent performance" } },
        { "match": { "review_text": "poor build quality" } }
      ]
    }
  }
}
```

In this query:
- We are using a bool query with two must clauses to ensure that both "excellent performance" and "poor build quality" are mentioned in the review text.
- The "review_text" field should be replaced with the actual field name in your dataset that contains the review text.

You can execute this query against your ElasticSearch database to get a list of reviews where the reviewer mentioned excellent performance but also complained about poor build quality for the same appliance.


In [56]:
example_milvus = f"""I have Amazon product reviews dataset. Use Milvus database system and create
code for the query: Get a list of those reviews that are similar to this text: “This
microwave started to make a lot of noise after using it for only few days” and got
reviewed in February of 2013."""

prompt = f"""
I have Amazon product reviews dataset. Use Milvus database system and create
code for the query: Get a list of those reviews that are similar to this text: Get a list of those reviews that are similar to this
text: Get a list of those reviews that are similar to this text: 
"The installation was a nightmare, took hours to figure out the connections and the manual was useless"
and were reviewed for range/oven products with ratings below 3 stars.

"""

response = get_completion(prompt)
print(response)


To achieve this, you can follow these steps:

1. Install Milvus database system and set it up.
2. Preprocess the Amazon product reviews dataset and create embeddings for each review text using a suitable text embedding model (e.g., BERT, Word2Vec, etc.).
3. Store the embeddings in Milvus database.
4. Define a function to calculate the similarity between the input text and each review text in the database.
5. Query the database to retrieve reviews that are similar to the input text and have ratings below 3 stars for range/oven products.

Here is a sample code snippet in Python using Milvus and Faiss for similarity search:

```python
import milvus
import faiss
import numpy as np

# Connect to Milvus
client = milvus.Milvus()

# Define the input text
input_text = "The installation was a nightmare, took hours to figure out the connections and the manual was useless"

# Preprocess the input text and create an embedding
# You can use a pre-trained text embedding model for this step

# Define 

In [57]:
example_milvus = f"""I have Amazon product reviews dataset. Use Milvus database system and create
code for the query: Get a list of those reviews that are similar to this text: “This
microwave started to make a lot of noise after using it for only few days” and got
reviewed in February of 2013."""

prompt = f"""
I have Amazon product reviews dataset. Use Milvus database system and create
code for the query: Get a list of those reviews that are similar to this text:
 "Great value for money but the build quality feels cheap and flimsy" and got reviewed between January 2012 and December 2013.
"""

response = get_completion(prompt)
print(response)

To achieve this, you can follow these steps:

1. Install Milvus database system and set it up.
2. Preprocess the Amazon product reviews dataset and create embeddings for each review text using a suitable text embedding model (e.g., BERT, Word2Vec, etc.).
3. Store the embeddings in Milvus database.
4. Use Milvus API to query for similar reviews based on the input text.
5. Filter the results based on the review date range (January 2012 to December 2013).

Here is a sample Python code snippet to perform the above steps:

```python
from milvus import Milvus, IndexType, MetricType
import numpy as np

# Connect to Milvus server
milvus = Milvus()

# Define the input text
input_text = "Great value for money but the build quality feels cheap and flimsy"

# Preprocess the input text and get its embedding
# You can use a pre-trained text embedding model for this step

# Query for similar reviews
query_embedding = np.array([input_text_embedding])  # Replace input_text_embedding with the actual emb

In [58]:
example_milvus = f"""I have Amazon product reviews dataset. Use Milvus database system and create
code for the query: Get a list of those reviews that are similar to this text: “This
microwave started to make a lot of noise after using it for only few days” and got
reviewed in February of 2013."""

prompt = f"""
I have Amazon product reviews dataset. Use Milvus database system and create
code for the query: Get a list of those reviews that are similar to this text:
Get a list of those reviews that are similar to this text: 
"The customer service was terrible, no help at all when I called about warranty issues" 
and are for products in the dishwasher category with helpful votes greater than 5.
"""

response = get_completion(prompt)
print(response)

To achieve this, you can follow these steps:

1. Install Milvus database system and set it up.
2. Preprocess the Amazon product reviews dataset and create embeddings for each review text using a suitable text embedding model (e.g., BERT, Word2Vec, etc.).
3. Store the embeddings in Milvus database.
4. Define a function to calculate the similarity between the input text and the review texts in the database.
5. Query the database to retrieve reviews that are similar to the input text and meet the specified criteria (i.e., for products in the dishwasher category with helpful votes greater than 5).

Here is a sample code snippet to perform the above steps:

```python
from milvus import Milvus, IndexType, MetricType
import numpy as np

# Connect to Milvus server
milvus = Milvus(host='localhost', port='19530')

# Define a function to calculate cosine similarity
def cosine_similarity(a, b):
    return np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b))

# Input text
input_text = "The custom

<div class="alert alert-block alert-danger">

***Requirement 5:**   (10 Points) Analyze, synthesize, and comment on the code
(Postgres SQL statements) generated by ChatGPT for every Quantitative
Query you identified above for completeness, correctness, and
consistency.

</div>

Query 1 SQL code:

Completeness: The code completely addresses the question. It will return exactly what was asked for:

The four specified percentiles (25th, 50th, 75th, 95th)
Calculated separately for each subcategory
Only for subcategories that have 10 or more products

Correctness: The PERCENT_RANK() calculation creates a column called percentile, but this column is never used in the main query so it makes it more complex than it should be. 

consistency: There are two quantative statements that may produce inconsistent readout due to NULL values. 
-- Step 1: Count for filtering (includes NULLs)
HAVING COUNT(*) >= 10

-- Step 2: Percentile calculation (may exclude NULLs)
PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY price)

Some edge cases may violate and return conflicting counts. 

Query 2:  Asks for Correlation between review length and helpfulness ratio
For each product subcategory With at least 50 reviews.

Completeness and Correctness of sql code: No it does not correcly identify based on product subcategory because the code uses .product_category. 
Additionally the question filters by at least 50 reviews but the code filters by HAVING COUNT(*) >= 50 product counts. 
Lastly, the code groups by product_category, product_id to do correlations but you can't do this because you need multiple observations. So correct way to do this is to
group by subcategories. So the code was not complete due to these errors.

Consistency: If vote count is 0 then helpful_votes / b.total_votes AS helpfulness_ratio may not work and generate inconsistent results. 
Additionally, -- Filtering counts all reviews including Null (HAVING COUNT(*) >= 50) but CORR(review_length, helpfulness_ratio) will only be done in complete observations data set generating inconsistent data use. 

Query 3: Question asks for: Brands with strongest seasonal performance variation Measured by highest difference between
 best and worst quarterly average ratingsFor brands with at least 100 reviews

Completeness and Correctness: There is a missing join statement "FROM quarterly_avg_ratings q1 ORDER BY seasonal_performance_variation DESC;" and
GROUP BY clause in the main query. This will query will include ALL brands, regardless of review count and filter is completely ignorred. 
Code also correctly filters brands with 100+ reviews in one section but it is never used. There are also critical syntax errors 
mixing aggregated and non-aggregated columns without GROUP BY statement. 

Consistency: Code will generate inconsistency with count and aggregated data outputs due to NULL value handling and there are temporal issues with 
extracting quarter information without year information. Q1 2022, Q1 2023, Q1 2024 are all treated as Quarter 1. 